commerce-taglib/src/main/java/com/liferay/commerce/taglib/servlet/taglib/SearchInputTag.java

```diff
@@ -0,0 +1,139 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.commerce.taglib.servlet.taglib;
+
+import com.liferay.commerce.taglib.servlet.taglib.internal.servlet.ServletContextUtil;
+import com.liferay.petra.string.CharPool;
+import com.liferay.petra.string.StringPool;
+import com.liferay.petra.string.StringUtil;
+import com.liferay.portal.kernel.util.HttpUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.taglib.util.IncludeTag;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.portlet.PortletURL;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.PageContext;
+
+/**
+ * @author Alessio Antonio Rendina
+ */
+public class SearchInputTag extends IncludeTag {
+
+	@Override
+	public int doStartTag() throws JspException {
+		_data = _getData();
+
+		return super.doStartTag();
+	}
+
+	public PortletURL getActionURL() {
+		return _actionURL;
+	}
+
+	public String getFormName() {
+		return _formName;
+	}
+
+	public void setActionURL(PortletURL actionURL) {
+		_actionURL = actionURL;
+	}
+
+	public void setFormName(String formName) {
+		_formName = formName;
+	}
+
+	@Override
+	public void setPageContext(PageContext pageContext) {
+		super.setPageContext(pageContext);
+
+		servletContext = ServletContextUtil.getServletContext();
+	}
+
+	@Override
+	protected void cleanUp() {
+		super.cleanUp();
+
+		_actionURL = null;
+		_data = null;
+		_formName = null;
+	}
+
+	@Override
+	protected String getPage() {
+		return _PAGE;
+	}
+
+	@Override
+	protected void setAttributes(HttpServletRequest httpServletRequest) {
+		request.setAttribute(
+			"liferay-commerce:search-input:actionURL", _actionURL);
+		request.setAttribute("liferay-commerce:search-input:data", _data);
+		request.setAttribute(
+			"liferay-commerce:search-input:formName", _formName);
+	}
+
+	private Map<String, Object> _getData() {
+		Map<String, Object> data = new HashMap<>();
+
+		_actionURL.setParameter("keywords", StringPool.BLANK);
+
+		String searchActionURL = String.valueOf(_actionURL);
+
+		List<String> parameters = StringUtil.split(
+			HttpUtil.getQueryString(searchActionURL), CharPool.AMPERSAND);
+
+		for (String parameter : parameters) {
+			if (parameter.length() == 0) {
+				continue;
+			}
+
+			List<String> parameterParts = StringUtil.split(
+				parameter, CharPool.EQUAL);
+
+			if (parameterParts.isEmpty()) {
+				continue;
+			}
+
+			String parameterValue = StringPool.BLANK;
+
+			if (parameterParts.size() > 1) {
+				parameterValue = parameterParts.get(1);
+			}
+
+			if (Validator.isNull(parameterValue)) {
+				continue;
+			}
+
+			parameterValue = HttpUtil.decodeURL(parameterValue);
+
+			data.put(parameterParts.get(0), parameterValue);
+		}
```

Comment:
It looks like this is mainly intended to parse a URL query string -- if that's true, we could probably replace a lot of this with `HttpUtil.getParameterMap`. A lot of the logic looks the same to me. And then we wouldn't have to maintain as much of this code.

Reply:
Hey Ethan, I did it like this to clean parameters without value

Reply:
Alright.... What would you think of using `HttpUtil.getParameterMap` and then removing the keys that don't have a value? That way we'd have to maintain less logic, and it'd be more obvious what/why we're doing.

Reply:
I think it's better this way instead of  HttpUtil.getParameterMap because the logic will be the same, but this way we clear the URL that is passed to the taglib that has different parameters compared to those from the request

Reply:
I don't understand. The logic would be much less the other way, I think?

	Map<String, String[]> parameterMap = HttpUtil.getParameterMap(searchActionURL);

	parameterMap.values().removeIf(
		param -> ArrayUtil.isEmpty(param) || ((param.length == 1) && Validator.isNull(param[0]))
	);

Reply:
This way doesn't work at all and even if it worked I had to add the logic to get value from the values array and set that parameter again into the data map. So it's pretty much the same amount of logic I think.

Reply:
Ok..... My problem with your code isn't that you're not doing it some particular way that I like; my problem is that this code is unreadable. No one on the team would be able to tell at a glance what it's trying to do, and this is _absolutely_ going to lead to maintainability issues. I'm just trying to make our code better. That's my job.

Why are we even parsing the URL string at all? We have a perfectly good `PortletURL` to make use of. How about this:

	private Map<String, Object> _getData() {
		Map<String, Object> data = new HashMap<>();

		_actionURL.setParameter("keywords", StringPool.BLANK);

		Map<String, String[]> parameterMap = _actionURL.getParameterMap();

		for (Map.Entry<String, String[]> parameterEntry : parameterMap.entrySet()) {
			String[] parameterValues = parameterEntry.getValue();

			if (ArrayUtil.isNotEmpty(parameterValues) &&
				Validator.isNotNull(parameterValues[0])) {

				data.put(parameterEntry.getKey(), parameterValues[0]);
			}
		}

		return data;
	}